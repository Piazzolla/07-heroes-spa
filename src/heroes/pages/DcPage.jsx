
export const DcPage = () => {
  return (
    <h1>Dc Page</h1>
  )
}
