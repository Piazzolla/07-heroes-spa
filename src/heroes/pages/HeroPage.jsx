import { useMemo } from "react";
import { Navigate, useNavigate, useParams } from "react-router-dom"
import { getHeroById } from "../helpers";

export const HeroPage = () => {

  const { id } = useParams();
  const hero = useMemo( () => getHeroById(id), [id]);
    /*uso memo para no llamar esto mas de una vez 
    (puede ser pesado si se conecta  una db etc)
    paso por parametro [id] y cuando cambie el id se vuelve a ejecutar
    la funcion del primer parametro
    */

  const navigate = useNavigate();
  const onNavigateBack = () => {
    if( hero.publisher === 'Marvel Comics')
      navigate('/marvel');
    else
      navigate('/dc')

    //si uso navigate(-1) vuelve a la ultima pagina que recuerda el navegador
  }

  if (!hero) {
    return <Navigate to="/marvel" />
  }

  return (
    <div className="row mt-5">
      <div className="col-4">
        <img
          src={`/assets/heroes/${ id }.jpg`}
          alt={ hero.superhero }
          className="img-thumbnail animate__animated animate__fadeInLeft"
        />
      </div>
      <div className="col-8">
        <h3>{hero.superhero}</h3>
        <ul className="list-group list-group-flush">
          <li className="list-group-item"> <b>Alter ego:</b> { hero.alter_ego } </li>
          <li className="list-group-item"> <b>Publisher:</b> { hero.publisher } </li>
          <li className="list-group-item"> <b>First appearance:</b> { hero.first_appearance } </li>
        </ul>

        <h5 className="mt-3">Characters</h5>
        <p>{ hero.characters} </p>

        <button className="btn btn-outline-primary" onClick={ onNavigateBack }>Regresar</button>
      </div>
    </div>
  )
}
