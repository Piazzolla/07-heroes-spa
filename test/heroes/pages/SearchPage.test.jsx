import { fireEvent, render, screen } from "@testing-library/react";
import { MemoryRouter } from 'react-router-dom';

import { SearchPage } from '../../../src/heroes/pages'



const mockedUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockedUseNavigate
}));

describe('Pruebas en <SearchPage />', () => {

    beforeEach( () => jest.clearAllMocks() );

    test('debe de mostrarse correctamente con valores', () => {
        const { container } = render(
            <MemoryRouter>
                <SearchPage />
            </MemoryRouter>
        )
        expect(container).toMatchSnapshot();
        //        screen.debug();

    });


    test('debe de mostrar a batman y el input con el valor del query', () => {
        render(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchPage />
            </MemoryRouter>
        );

        const input = screen.getByRole('textbox');
        expect( input.value ).toBe('batman');
        const img = screen.getByRole('img');
        expect( img.src ).toContain('/assets/heroes/dc-batman.jpg');
        const searchDiv = screen.getByLabelText('search');
        expect( searchDiv.style.display ).toBe('none');

    });

    test('debe mostrar un error si no se encuentra el hero (batman123)', () => {
        render(
            <MemoryRouter initialEntries={['/search?q=batman123']}>
                <SearchPage />
            </MemoryRouter>
        );

        const searchDiv = screen.getByLabelText('hero-error');
        expect( searchDiv.style.display ).not.toBe('none');
      })

    test('debe llamar el navigate en la pantalla nueva', () => { 
        render(
            <MemoryRouter initialEntries={['/search']}>
                <SearchPage />
            </MemoryRouter>
        );


        

        const input = screen.getByRole('textbox');
        fireEvent.change(input, { target: { name: 'searchText', value: 'batman' }})
        const form = screen.getByLabelText('form');
        fireEvent.submit( form );
        expect ( mockedUseNavigate ).toHaveBeenCalledWith(`?q=batman`); 
     })
});