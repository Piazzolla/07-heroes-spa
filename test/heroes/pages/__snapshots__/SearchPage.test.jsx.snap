// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchPage /> debe de mostrarse correctamente con valores 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search a hero"
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-1"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        aria-label="search"
        class="alert alert-primary animate__animated animate__fadeIn"
      >
        Search a hero
      </div>
      <div
        aria-label="hero-error"
        class="alert alert-danger animate__animated animate__fadeIn"
        style="display: none;"
      >
        No hero with 
        <b />
      </div>
    </div>
  </div>
</div>
`;
